lib/std/builtin.zig

```diff
@@ -792,10 +792,10 @@ pub fn default_panic(msg: []const u8, error_return_trace: ?*StackTrace, ret_addr
             var exit_data = ExitData.create_exit_data(msg, &exit_size) catch null;
 
             if (exit_data) |data| {
-                if (uefi.system_table.std_err) |out| {
+                if (uefi.system_table.con_out) |out| {
                     _ = out.setAttribute(uefi.protocols.SimpleTextOutputProtocol.red);
                     _ = out.outputString(data);
-                    _ = out.setAttribute(uefi.protocols.SimpleTextOutputProtocol.white);
+                    _ = out.setAttribute(uefi.protocols.SimpleTextOutputProtocol.lightgray);
                 }
             }
 
```

Comment:
CC @fifty-six you probably have an opinion on this (mainly the `std_err` -> `con_out` change). If you do agree with this change though, do you think we can also get rid of the leading `\r\n` or have you had a case where that would cause the error to not be on a new line? If you look at the screenshot in the description of the relevant commit https://github.com/ziglang/zig/pull/16094/commits/76732ae50799a4910939b102c77f9666eae46109 you'll see an extra, empty line.

Reply:
If anything, it might be a good idea to do both - I forgot to do so in `builtin.zig`, but that's [what I was, and have been, doing](https://github.com/fifty-six/zbl/blob/master/src/panic.zig#L52) for my own project before I pr'd the change to builtin.zig (and I forgot to swap over to that after, whoops). QEMU (using OMVF)-wise, `con_out` goes the screen and the attached serial ports, whereas `stderr` is only to the serial ports, which I've found useful in cases - you can see this in the image below also. 

In retrospect, if we're only doing one - `con_out` is *definitely* a better choice, as I forgot you have to specify `-serial stdio` in order to see that for QEMU, and you don't really see `stderr` when booting off an actual disk.

Regarding the leading `\r\n`, this can make a difference if the cursor is set to a certain position - I hit a case with this while making a tui-style menu. 
![err: before some previously written output in a line. cutting into it](https://github.com/ziglang/zig/assets/10470872/7480d9c5-c666-447b-a5a3-01f414044666)
The `\r\n` there makes it so it's at the start of a new line, but it does have the downside of the blank line in other cases. I think it's fine to remove if you think that's a good idea though.

For what it's worth, taking a look at uefi-rs, they [just print to con_out](https://github.com/rust-osdev/uefi-rs/blob/163aea3be8d71a4a0a503e465dc993ea377d903d/uefi-services/src/lib.rs#L201), though they also special case QEMU.

Reply:
Ah, I see! Yes, it might be good if we print it to both stdout and the serial stderr then.

Technically it'd be more consistent to omit the leading CRLF because the cursor can be changed in for example a terminal on Linux too, using escape codes for example. So, it would be more consistent to either have a leading newline like that on all platforms or on none. The `std.log.err("{s}", .{@errorName(err)});` in start.zig doesn't have a trailing newline. I think for now I'm going to remove it.
And eventually maybe we can use `std.log.err` on the UEFI too once the `os.zig` abstractions advance and then maybe we're going to use the same code.

Yes, I was thinking the same. In my cases where I tested it in QEMU I have the luck that it just stays there with the error message on screen more or less but on real hardware it might be different.
I think we could even do like `shutting down after input...` and wait until there's input (or just without that extra message).
I think this will ultimately just result in a better user experience. I'm noticing that the code in builtin.zig and start.zig will be very similar and I wonder if we can use the same code in both cases.

The main difference is that you heap-allocate the utf16 buffer rather than using stack memory. Any reason for that? My code in start.zig is using stack memory and it seems to have worked fine.

Another difference is `_ = bs.exit(uefi.handle, .Aborted, exit_size, exit_data);` whereas in start.zig it just returns the `std.os.uefi.Status`. I was curious what exactly happens with the `exit_size` and `exit_data` when passed to `bs.exit`?

Reply:
Removing `\r\n` sounds good to me then!

I think waiting for input is a good idea for the same reason as the justification we could have for not having it, it's super easy to replace the panic handler - and it's definitely more user friendly for us to default to waiting. If you want to see how to do it (or I could PR it if you'd like), I have a decent impl in my panic handler.

Yeah, I was thinking maybe we want to stack allocate it first, and then copy it over to the heap - in the case that the pool allocator is for some reason unavailable or there's no memory left. That way, we can still display the error message, even if we can't pass it to exit itself. 

Regarding ExitData, you can see its usage in `EFI_BOOT_SERVICE`'s `StartImage`, one example of usage would be chainloading - if my bootloader tried to start another one, but it panic'd, I'd get the exit data back and be able to display that / do whatever I want with the given information.

Discarding the status is so we have an option in the case of a failure to shut down, I believe most impls of similar things just stall if they can't shutdown properly, though here I just fallback to whatever else was implemented.

Reply:
Of course! If you want you can make a PR for that. That might just be better so that I don't get carried away too much here...

Using the stack first and then the heap is a great idea by the way. Actually, that might just be what I need to merge the two implementations!

Reply:
@fifty-six I actually got one more question: we seem to be passing either 256 or 0 as the exit_size. Why not use the actual length of the produced error message?

Reply:
I just went for the size of the allocated buffer, which is statically 256 — though it might be better to have some estimate based off the message length. I believe we could go with either of the options (though for length, we'd want the utf16 length including the terminator), as the pool allocator free doesn't need the initial size of the allocation.